utils/bundle_miner.c

```diff
@@ -0,0 +1,157 @@
+      if (probability < ctx->probability) {
+        ctx->probability = probability;
+        ctx->optimal_index = ctx->index;
+      }
+    }
+
+    ctx->index += 1;
+  }
+
+  return NULL;
+}
+
+/**
+ * Public functions
+ */
+
+double bundle_miner_probability_of_losing(byte_t const *const normalized_hash, uint8_t const security) {
+  double p = 0.0;
+  double pi = 0.0;
+
+  for (size_t i = 0; i < security * NORMALIZED_FRAGMENT_LENGTH; i++) {
+    pi = 1.0 - ((double)(TRYTE_VALUE_MAX - normalized_hash[i]) / (double)(TRYTE_VALUE_MAX - TRYTE_VALUE_MIN));
+    if (pi > 0.0) {
+      if (p == 0.0) {
+        p = 1.0;
+      }
+      p *= pi;
+    }
+  }
+
+  return p;
+}
+
+void bundle_miner_normalized_bundle_max(byte_t const *const lhs, byte_t const *const rhs, byte_t *const max,
+                                        size_t const length) {
+  for (size_t i = 0; i < length; i++) {
+    max[i] = MAX(lhs[i], rhs[i]);
+  }
+}
+
+retcode_t bundle_miner_mine(byte_t const *const bundle_normalized_max, uint8_t const security,
+                            trit_t const *const essence, size_t const essence_length, uint32_t const count,
+                            uint8_t const nprocs, uint64_t *const index) {
+  retcode_t ret = RC_OK;
+  uint8_t procs = nprocs == 0 ? system_cpu_available() : nprocs;
+  uint32_t rounded_count = count + (procs - count % procs);
+  bundle_miner_ctx_t *ctxs = (bundle_miner_ctx_t *)malloc(sizeof(bundle_miner_ctx_t) * procs);
+  thread_handle_t *threads = (thread_handle_t *)malloc(sizeof(thread_handle_t) * procs);
+  uint64_t start_index = 0;
+  double probability = 1.0;
+
+  if (bundle_normalized_max == NULL || essence == NULL || index == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  if (security > 3 || essence_length % HASH_LENGTH_TRIT != 0) {
+    return RC_UTILS_BUNDLE_MINER_BAD_PARAM;
+  }
+
+  if (ctxs == NULL || threads == NULL) {
+    return RC_OOM;
+  }
+
+  start_index = trits_to_long(essence + OBSOLETE_TAG_OFFSET, NUM_TRITS_OBSOLETE_TAG);
+  *index = 0;
+
+  for (size_t i = 0; i < procs; i++) {
+    ctxs[i].bundle_normalized_max = bundle_normalized_max;
+    ctxs[i].security = security;
+    if ((ctxs[i].essence = (trit_t *)malloc(sizeof(trit_t) * essence_length)) == NULL) {
+      ret = RC_OOM;
+      goto done;
+    }
+    memcpy(ctxs[i].essence, essence, essence_length);
+    ctxs[i].essence_length = essence_length;
+    ctxs[i].index = start_index + (i * (rounded_count / procs));
+    ctxs[i].count = rounded_count / procs;
+    ctxs[i].optimal_index = 0;
+    ctxs[i].probability = 1.0;
+    thread_handle_create(&threads[i], (thread_routine_t)bundle_miner_mine_routine, &ctxs[i]);
+  }
+
+  for (size_t i = 0; i < procs; i++) {
+    thread_handle_join(threads[i], NULL);
+    if (ctxs[i].probability < probability) {
+      probability = ctxs[i].probability;
+      *index = ctxs[i].optimal_index;
+    }
+    free(ctxs[i].essence);
+  }
+
+done:
+
+  free(ctxs);
```

Comment:
what if some of the ctxs have `essence` allocated?

Reply:
Right